lib/semantics/semantics.cc

```diff
@@ -67,10 +67,140 @@ template<typename... C> class SemanticsVisitor : public virtual C... {
     Enter(node);
     return true;
   }
+  template<typename T> bool Pre(const parser::UnlabeledStatement<T> &node) {
+    context_.set_location(node.source);
+    Enter(node);
+    return true;
+  }
   template<typename T> void Post(const parser::Statement<T> &node) {
     Leave(node);
     context_.set_location(std::nullopt);
   }
+  template<typename T> void Post(const parser::UnlabeledStatement<T> &node) {
+    Leave(node);
+    context_.set_location(std::nullopt);
+  }
+
+  bool Pre(const parser::AssociateConstruct &associateConstruct) {
```

Comment:
It would be great if you could combine all of these repetitive `Pre` overloads.  It's hard to see the differences between them, if any, and it's hard to tell which ones are paired up with `Post` overloads and which are not.

Reply:
Specifically, you could augment the default `Pre()` and `Post()` member function templates above (ca. line #60) with something like

'''
if constexpr (common::HasMember<const N *, ConstructNode>) {
  context_.PushConstruct(node);
}
```

Reply:
Peter, you're a genius!  I wanted to combine those calls, but I couldn't figure out how to do it.